x-pack/platform/plugins/shared/alerting/server/maintenance_window_events/task.ts

```diff
@@ -0,0 +1,284 @@
+    ),
+    nodeBuilder.range('maintenance-window.attributes.events', 'gte', 'now'),
+  ]);
+  const finished = nodeBuilder.and([
+    nodeTypes.function.buildNode(
+      'not',
+      nodeBuilder.range('maintenance-window.attributes.events', 'gte', 'now')
+    ),
+    nodeBuilder.range('maintenance-window.attributes.expirationDate', 'gte', 'now'),
+  ]);
+
+  return nodeBuilder.or([running, upcoming, finished]);
+}
+
+export async function findMaintenanceWindowsSo({
+  savedObjectsClient,
+  logger,
+  filter,
+}: {
+  logger: Logger;
+  savedObjectsClient: ISavedObjectsRepository;
+  filter: KueryNode;
+}) {
+  try {
+    const finder = await savedObjectsClient.createPointInTimeFinder<MaintenanceWindowAttributes>({
+      type: MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE,
+      namespaces: ['*'],
+      perPage: 1000,
+      sortField: 'updatedAt',
+      sortOrder: 'desc',
+      filter,
+    });
+
+    let result: Array<SavedObject<MaintenanceWindowAttributes>> = [];
+    for await (const findResults of finder.find()) {
+      result = result.concat(findResults.saved_objects);
+    }
+
+    await finder.close();
+
+    return result;
+  } catch (e) {
+    logger.error(`Failed to find maintenance windows saved object". Error: ${e.message}`);
+    return [];
+  }
+}
+
+interface GenerateEventsAndUpdateSO {
+  logger: Logger;
+  savedObjectsClient: ISavedObjectsRepository;
+  maintenanceWindowsSO: Array<SavedObject<MaintenanceWindowAttributes>>;
+  startRangeDate: string;
+}
+
+export async function generateEventsAndUpdateMaintenanceWindowSavedObjects({
+  logger,
+  savedObjectsClient,
+  maintenanceWindowsSO,
+  startRangeDate,
+}: GenerateEventsAndUpdateSO) {
+  let updatedMaintenanceWindows: Array<SavedObjectsUpdateResponse<MaintenanceWindowAttributes>> =
+    [];
+  const maintenanceWindows = maintenanceWindowsSO.map((savedObject) =>
+    transformMaintenanceWindowAttributesToMaintenanceWindow({
+      attributes: savedObject.attributes,
+      id: savedObject.id,
+    })
+  );
+
+  const newExpirationDate = moment(startRangeDate).utc().add(1, 'year').toISOString();
+
+  const mwWithGeneratedEvents = maintenanceWindows
+    .filter(
+      (maintenanceWindow) =>
+        (maintenanceWindow.rRule.interval !== undefined ||
+          maintenanceWindow.rRule.freq !== undefined) &&
+        maintenanceWindow.events.length
+    )
+    .map((filteredMaintenanceWindow) => {
+      const {
+        events: oldEvents,
+        rRule,
+        duration,
+        expirationDate: oldExpirationDate,
+      } = filteredMaintenanceWindow;
+
+      const newEvents = generateMaintenanceWindowEvents({
+        rRule,
+        expirationDate: newExpirationDate,
+        duration,
+        startDate: oldExpirationDate,
+      });
+
+      return {
+        ...filteredMaintenanceWindow,
+        expirationDate: newEvents.length ? newExpirationDate : oldExpirationDate,
+        events: [...oldEvents, ...newEvents],
```

Comment:
Do we need to keep the old events? I am wondering a) if the old events become so big that they cannot fit in memory, and b) if the events array becomes too big and the ES grows too much. The new events can contain some past events to be sure we do not miss anything. We can do that by setting the `startDate` of the `generateMaintenanceWindowEvents` to be one week in the past.

Reply:
`startDateRange` used in filter is 1 week before expiration date, hence used the same date to generate new events.